geode-membership/src/main/java/org/apache/geode/distributed/internal/membership/gms/MemberIdentifierImpl.java

```diff
@@ -693,10 +701,43 @@ public void fromData(DataInput in,
         // nope - it's from a pre-GEODE client or WAN site
       }
     }
+    convertIpToHostnameIfNeeded();
+  }
+
+  public void fromDataPre_GEODE_1_15_0_0(DataInput in, DeserializationContext context)
+      throws IOException, ClassNotFoundException {
+    readMemberData(in);
+    memberData.readAdditionalData(in);
+    convertIpToHostnameIfNeeded();
   }
 
   public void fromDataPre_GFE_9_0_0_0(DataInput in, DeserializationContext context)
       throws IOException {
+    readMemberData(in);
+    convertIpToHostnameIfNeeded();
+  }
+
+  /**
+   * In older versions of Geode {@link MemberData#getHostName()} can return an IP address
+   * if network partition is enabled. Also, older versions of Geode don't use bind-address for
+   * SNI endpoint identification and do a reverse lookup to find the fully qualified hostname.
+   * This can become a problem if TLS certificate doesn't have the fully qualified name in it
+   * as a Subject Alternate Name, therefore this behavior was changed to preserve the bind-address.
```

Comment:
Second sentence confuses me. What do you think of this:

In older versions of Geode {@link MemberData#getHostName()} can return an IP address if network partition detection is enabled. If SSL endpoint identification is enabled, those product versions supply the result of a reverse lookup to the TLS handshake API. Endpoint identification will fail if e.g. the lookup returned a fully-qualified name but the certificate had just a (non-fully-qualified) hostname in a Subject Alternate Name field.

In version 1.15.0 member identifiers were changed so that if a bind address is specified, that exact string will be carried as the host name. That gives the administrator better control over endpoint identification. When upgrading from earlier versions we convert any IP numbers to hostnames via reverse lookup here.